### PR TITLE
Fix App path for usability.

### PR DESCRIPTION
**Pull request by dereuromark**
Resolves problematic usability described in https://github.com/cakephp/cakephp/issues/13604

- `App::path()` remains for all non class path elements.
- `App::classPath()` has been added and can proxy Plugin::classPath()
- App::path() is allowed to also use plugin, same as others for convenience.

Does that make sense?

The changes mean that e.g. my IdeHelper can keep the simple App::className($type, $plugin) syntax here for simplicity.

I wasn't sure if we should extract also templatePath(), since locales and other such keys don't have this yet either.
Or do we plan to add `public function getLocalePath()` to Plugin class as well?

### CODE REVIEW

File: src/Core/App.php [hunk 1]
```diff
@@ -179,31 +179,45 @@ protected static function _classExistsInBase(string $name, string $namespace): b
...
-            return [APP . $type . DIRECTORY_SEPARATOR];
+        if ($type === 'templates') {
+            return [Plugin::templatePath($plugin)];
         }
 
         deprecationWarning(
-            'Using App::path() with 2nd argument $plugin is deprecated.'
-            . ' Use \Cake\Core\Plugin::classPath()/templatePath() instead.'
+            'App::path() is deprecated for class path.'
+            . ' Use \Cake\Core\App::classPath() instead or directly the method on \Cake\Core\Plugin.'
```
Comment (markstory) [inline src/Core/App.php]:
```suggestion
            . ' Use \Cake\Core\App::classPath() or \Cake\Core\Plugin::classPath() instead.'
```

File: src/Core/App.php [hunk 2]
```diff
@@ -179,31 +179,45 @@ protected static function _classExistsInBase(string $name, string $namespace): b
...
         );
 
-        if ($type === 'templates') {
-            return [Plugin::templatePath($plugin)];
+        return static::classPath($type, $plugin);
+    }
+
+    /**
```
Comment (markstory) [inline src/Core/App.php]:
```suggestion
    /**
     * Get the path to a class type in the application or a plugin.
     *
```

File: src/Core/App.php [hunk 3]
```diff
@@ -179,31 +179,45 @@ protected static function _classExistsInBase(string $name, string $namespace): b
      * Will return the path for tables `src/Model/Table`.
      *
      * @param string $type type of path
-     * @param string|null $plugin This argument is deprecated.
-     *   Use \Cake\Core\Plugin::classPath()/templatePath() instead for plugin paths.
-     * @return array
+     * @param string|null $plugin This argument is deprecated for class paths.
+     *   Use \Cake\Core\Plugin::classPath() instead or directly the method on \Cake\Core\Plugin.
+     * @return string[]
+     * @deprecated 4.0 Use \Cake\Core\App::classPath()/templatePath() instead.
```
Comment (ADmad) [inline src/Core/App.php]: But there's not `App::templatePath()` currently.

Comment (dereuromark) [reply src/Core/App.php]: Do we want to create it? Or rather keep all lowercase "templates"/"locales" here in App::path()?

File: src/Core/App.php [hunk 4]
```diff
@@ -178,32 +178,63 @@ protected static function _classExistsInBase(string $name, string $namespace): b
...
+     * Example:
+     *
+     * ```
+     * App::path('Model/Table');
+     * ```
+     *
+     * Will return the path for tables `src/Model/Table`.
+     *
+     * ```
+     * App::path('My/Plugin.Model/Table');
```
Comment (ADmad) [inline src/Core/App.php]: Method name as well the argument doesn't seem right.

Comment (dereuromark) [reply src/Core/App.php]: Copy and paste ;) thx!

**Comment by dereuromark** [comment]
I also added a note about default types for path(), since this is otherwise hard to retrieve from code.

File: src/Core/App.php [hunk 5]
```diff
@@ -172,38 +177,63 @@ protected static function _classExistsInBase(string $name, string $namespace): b
...
+     * Example:
+     *
+     * ```
+     * App::classPath('Model/Table');
+     * ```
+     *
+     * Will return the path for tables `src/Model/Table`.
+     *
+     * ```
+     * App::classPath('My/Plugin.Model/Table');
```
Comment (ADmad) [inline src/Core/App.php]: The argument still doesn't seem right to me. Shouldn't it be `App::classPath('Model/Table', 'My/Plugin')`?

Comment (dereuromark) [reply src/Core/App.php]: I guess my brain was too much into Router::path()^^

File: tests/TestCase/Core/AppTest.php [hunk 6]
```diff
@@ -216,6 +216,22 @@ public function shortNameProvider()
...
+     * test classPath() with a plugin.
+     *
+     * @return void
+     */
+    public function testClassPathWithPlugins()
+    {
+        $basepath = TEST_APP . 'Plugin' . DS;
+        $this->loadPlugins(['TestPlugin', 'Company/TestPluginThree']);
+        $result = App::classPath('Controller', 'TestPlugin');
+        $this->assertPathEquals($basepath . 'TestPlugin' . DS . 'src' . DS . 'Controller' . DS, $result[0]);
```
Comment (ADmad) [inline tests/TestCase/Core/AppTest.php]: Nitpick: A newline after this assert would have been nice for easier reading :slightly_smiling_face:.

Comment (dereuromark) [reply tests/TestCase/Core/AppTest.php]: True. I just copied them from the other test case.

**Review by ADmad — APPROVED** [review]
Changes LGTM. The plugins would also have to be updated to use the new `classPath()` method instead of `path()` wherever applicable.

**Comment by dereuromark** [comment]
Docs: https://github.com/cakephp/docs/pull/6279

**Pull request merged**